Parse arbitrary tabla bol and return list of stroke sensities for each matra

amoghpj opened the pull request:
Implements discussion in #7

amoghpj committed:
Add bolparser and taals.json

amoghpj committed:
Rename Akshar to Matra, start working on tokeninzing

amoghpj committed:
Functional tokenization

amoghpj committed:
Test general tokenization for n avartans

amoghpj committed:
Tested

amoghpj committed:
Rename function

amoghpj commented:
The tokenization is pretty ad hoc. This worked well for the bols I came up with, but it has 
to be tested more rigorously. It will work fine for straightforward tukdas though.

reckoner165 commented on bolparser.py:
```diff
@@ -0,0 +1,118 @@
+class bolParser:
+    def __init__(self, bol):
+        self.bol = bol
+        self.Avartan = list()
```
Either up here, or by the getter functions, can you add a comment each defining `Avartan`, `Vibhaag` and `Matra`?

amoghpj replied:
Updated this in the most recent commit.

reckoner165 commented on bolparser.py:
```diff
@@ -0,0 +1,118 @@
+
+    def __getMatra(self):
+        GroupedAkshars = [[v.split(' ') for v in V] for V in self.Vibhaag]
+        for av in GroupedAkshars:
```
What's a, v and av in this case?

amoghpj replied:
v for vibhaag, av for avartan. The `a` was a remnant of an older convention, updated it to `m` to stand for matra now.
Do you have a different convention I can follow?

reckoner165 commented on bolparser.py:
```diff
@@ -0,0 +1,118 @@
+                loc = matra.find(C)
+                if loc != -1:
+                    matra = matra.replace(C,'',1)
+                    if C == 'tirakita':
```
I think this could be a dict where each compound token is mapped to a number

amoghpj replied:
I updated this in the latest commit as well.

reckoner165 commented on bolparser.py:
```diff
@@ -0,0 +1,118 @@
+                    print("Unidentified akshar!\n Dump:")
+                    print(matra)
+                    print("Exiting..")
+                    sys.exit()
```
Better to avoid `sys.exit` as we need the main REPL running regardless. 
This should probably `pass` and wait for the next bol/input. 
Maybe call a reset method so that all variables in this class get flushed.

amoghpj replied:
Good idea. I'm not familiar with the `code` package. Can you point me to  some documentation on the correct way to handle an exception?

reckoner165 replied:
I think raising a `ValueError` should be good. This will also ensure the user doesn't get kicked out of the REPL.
http://hplgit.github.io/primer.html/doc/pub/input/._input-bootstrap007.html

reckoner165 commented on bolparser.py:
```diff
@@ -0,0 +1,118 @@
+            count = 0
+            while len(matra) > 0:
+                valid = False
+                for cs in alltokens:
```
is this a check to ensure everything in the matra is accounted for?

amoghpj replied:
Yes, this prevents getting stuck in an infinite loop if there is something undefined in the string.

reckoner165 commented on bolparser.py:
```diff
@@ -0,0 +1,118 @@
+            matraDensity.append(ad)
+        return(matraDensity)
+
+    # TEST
```
Worth turning these into unit tests so we don't break the parser inadvertently in future. What do you think?

amoghpj replied:
I was thinking that as well! Just that I don't have experience writing unit tests having never written one. I found doctest, would you recommend using this? Also, will there be a single file containing all the tests?

amoghpj committed:
Update in response to comments on PR #10

1. Add documentation defining Matra, Vibhaag, and Avartan with
   example
2. Update variable name from 'a' to 'm' in __getMatra()
3. Created dictionary compoundTokenDensity to replace
   conditional statements

amoghpj committed:
new gitignore

amoghpj committed:
Merge remote-tracking branch 'upstream/master'

amoghpj committed:
Add test functions to attempt a meend

1. Add meend() in repl.py
2. Modify status
3. Add else clause in play() to check the key 'type' in status in
   sequencer.py
4. Add glide() to Modular, which currently creates a linear
   interpolation or a non-linear smooth step interpolation between
   pairs of notes

amoghpj commented:
Latest commit implements a glide() function in soundmodular

reckoner165 commented on repl.py:
```diff
@@ -36,8 +37,14 @@ def avaroha(num_notes, stutter=1):
+    play_notes = raag['avaroha']
+    sequence_notes(play_notes, num_notes, stutter, 'glide')
+    
+def sequence_notes(note_list, num_notes, stutter, outtype='osc'):
```
thoughts on making `glide` a boolean?

reckoner165 commented on packages/soundmodular.py:
```diff
@@ -369,3 +413,23 @@ def silence(self, duration):
         return out_block
+    
+    def fit_transition(self, low, high, num_points):
+        print(low)
```
take away print statements?

reckoner165 commented on packages/soundmodular.py:
```diff
@@ -369,3 +413,23 @@ def silence(self, duration):
 
         return out_block
+    
+    def fit_transition(self, low, high, num_points):
```
For posterity, can you add a comment on what this method does?

reckoner165 commented on packages/soundmodular.py:
```diff
@@ -210,6 +210,50 @@ def osc_tone(self, duration, frequency):
+            currfreq = start_freq
+            fs = [start_freq+i*fstep for i in range(len(times))]
+            # Quadratic
+            fs = self.fit_transition(start_freq, end_freq, num_samples_per_split)
```
Worth trying out logarithmic, since humans perceive pitch changes on a log (base 2) scale
